### PR TITLE
Added combine storage with cache lifetime and mulitple storage modes

### PR DESCRIPTION
**Pull request by th-mustache-dk**

### CODE REVIEW

**Review by kasperhansen — APPROVED** [review]

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+    }
+    
+    private func isStillValid(cachedAt: Date) -> Bool {
+        
+        guard let expiration = self.expiration else { return true }
+        
+        switch expiration {
+            case .seconds(let seconds):
+                let expirationTime = cachedAt.addingTimeInterval(seconds)
+                if expirationTime > .nowSafe {
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: Hmm, intuitively I would think that expirationTime is before now when value is invalid...

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+                return cache.value
+        }
+    }
+    
+    func setMemory(scope: MemoryScope, value: T?) {
+        switch scope {
+            case .singleton:
+                
+                let key = String(describing: T.self)
+                singletonMemoryContainer[key] = value
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: Should be wrapped in CacheContainer - to have the expire date and match the getter above.

File: Sources/MustacheFoundation/Classes/Extensions/String+Extensions.swift [hunk 3]
```diff
@@ -17,14 +17,16 @@ public extension String {
...
+    } 
     
     func stripOutHtml() -> String? {
         do {
             guard let data = self.data(using: .unicode) else {
                 return nil
             }
-            let attributed = try NSAttributedString(data: data, options: [.documentType: NSAttributedString.DocumentType.html, .characterEncoding: String.Encoding.utf8.rawValue], documentAttributes: nil)
+            let attributed = try NSAttributedString(data: data,
+                                                    options: [.documentType: NSAttributedString.DocumentType.html, .characterEncoding: String.Encoding.utf8.rawValue],
```
Comment (kasperhansen) [inline Sources/MustacheFoundation/Classes/Extensions/String+Extensions.swift]: Maybe 'let options' on a separate line instead of line breaking.

File: Tests/MustacheCombineTest/StorageDefaultsTest.swift [hunk 4]
```diff
@@ -0,0 +1,93 @@
+
+import XCTest
+import MustacheCombine
+
+@available(iOS 13.0, *)
+final class StorageDefaultsTest: XCTestCase {
+    
+    @StorageCombine("memorySingletonNone", mode: .userDefaults(), expiration: .seconds(2))
```
Comment (kasperhansen) [inline Tests/MustacheCombineTest/StorageDefaultsTest.swift]: Annotation key should match the variable name.

File: Tests/MustacheCombineTest/StorageDefaultsTest.swift [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        
+        let stored = StoredObject()
+        
+        XCTAssertNil(self.defaultsDay)
+        self.defaultsDay = stored
+        XCTAssertEqual(self.defaultsDay, stored)
+        
+    }
+    
+    func testDefaultsDate() throws {
```
Comment (kasperhansen) [inline Tests/MustacheCombineTest/StorageDefaultsTest.swift]: Maybe use the naming 'defaultsNow' for this testcase/property - and move it to be the first testcase/property.
We could also add a test for negative expiration.

File: Tests/MustacheCombineTest/StorageMemoryTest.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    var memoryUniqueNone: StoredObject?
+    
+    override func setUpWithError() throws {
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+        self.memorySharedNone = nil
+        XCTAssertNil(self.memorySharedNone)
```
Comment (kasperhansen) [inline Tests/MustacheCombineTest/StorageMemoryTest.swift]: We should probably not assert in teardown.

File: Tests/MustacheCombineTest/StorageMemoryTest.swift [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        
+        XCTAssertNil(self.memorySingletonNone)
+        self.memorySingletonNone = stored
+        XCTAssertEqual(self.memorySingletonNone, stored)
+                
+    }
+
+    func testMemorySingletonNone2() throws {
+        
+        XCTAssertNotNil(self.memorySingletonNone)
```
Comment (kasperhansen) [inline Tests/MustacheCombineTest/StorageMemoryTest.swift]: We cannot trust the order of tests, so we should remove or rethink this one.
Maybe add another property and without setting it check the object is the same (if that's the case?)

File: Tests/MustacheCombineTest/StorageMemoryTest.swift [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    }
+    
+    func testMemoryUniqueNone() throws {
+        
+        let stored = StoredObject()
+        
+        XCTAssertNil(self.memoryUniqueNone)
+        self.memoryUniqueNone = stored
+        
+        XCTAssertEqual(self.memoryUniqueNone, stored)
```
Comment (kasperhansen) [inline Tests/MustacheCombineTest/StorageMemoryTest.swift:65]: We should test that 'memorySharedNone' is Nil (not related), since that's the special case here.

**Review by kasperhansen — COMMENTED** [review]
Some small style and test comments - I'll review the main file next.

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+                KeychainWrapper.standard.removeObject(forKey: self.key, withAccessibility: accessibility)
+            case .memory(let scope):
+                switch scope {
+                    case .singleton:
+                        let key = String(describing: T.self)
+                        singletonMemoryContainer[key] = nil
+                    case .unique:
+                        self.uniqueMemoryStorage = nil
+                    case .shared:
+                        sharedMemoryValueContainer.removeObject(forKey: self.sharedMemoryKey)
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift:182]: Do we also need to remove it from the keyContainer? Or will it break other references...

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+            self.clear()
+            return nil
+        }
+        
+        return cache.value
+    }
+    
+    func setUserDefaults(defaults: UserDefaults, value: T?, createdAt: Date = .nowSafe) {
+        if let value = value {
+            let cache = CacheContainer(value: value, createdAt: Date())
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift:227]: Use the provided 'createdAt' parameter for CacheContainer (or remove it).

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+        
+        switch self.mode {
+            case .userDefaults(let defaults):
+                defaults.removeObject(forKey: self.key)
+            case .keychain(let accessibility):
+                KeychainWrapper.standard.removeObject(forKey: self.key, withAccessibility: accessibility)
+            case .memory(let scope):
+                switch scope {
+                    case .singleton:
+                        let key = String(describing: T.self)
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: I think we need self.key here?

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+
+// MARK: StorageMode.memory
+@available(iOS 13.0, *)
+extension StorageCombine {
+    
+    func getMemory(scope: MemoryScope) -> T? {
+        switch scope {
+            case .singleton:
+            
+                guard let cache = singletonMemoryContainer[self.key] as? CacheContainer<T?>, let value = cache.value else { return nil }
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: We probably don't need to guard for value - although it makes no practical difference.

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+                    return nil
+                }
+                
+                return cache.value
+                
+            case .shared:
+
+                guard let cache = sharedMemoryValueContainer.object(forKey: self.sharedMemoryKey) as? CacheContainer<T> else { return nil }
+                
+                guard self.isStillValid(cachedAt: cache.createdAt) else {
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: It would be nice to de-duplicate the cache check by moving it outside the switch.
Prepare the variable before the switch: `let cache: CacheContainer<T?>?` then each case could be a single line assigning it.

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+                if let value = value {
+                    let cache = CacheContainer(value: value, createdAt: Date())
+                    sharedMemoryValueContainer.setObject(cache, forKey: self.sharedMemoryKey)
+                } else {
+                    sharedMemoryValueContainer.removeObject(forKey: self.sharedMemoryKey)
+                }
+                
+                // Sends notification so that subscribers of CurrentValueSubject gets the newest object
+                NotificationCenter.default.post(name: notificationName(key: self.key),
+                                                object: self,
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: Probably notify with nil object in this case.

File: Sources/MustacheCombine/MustacheServices/StorageCombine.swift [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+        self.value = value
+        self.createdAt = createdAt
+    }
+    
+    static func < (lhs: CacheContainer, rhs: CacheContainer) -> Bool {
+        return lhs.createdAt < rhs.createdAt
+    }
+}
+
+public extension Double {
```
Comment (kasperhansen) [inline Sources/MustacheCombine/MustacheServices/StorageCombine.swift]: Could be moved to Numbers+Extension or similar.
